File: recursion/binary-subtree/javascript/BinarySubtree.js
class Node{
    constructor(data){
        this.data = data;
        this.left = null;
        this.right = null;
    }
}


//             1
//           2    3
//         4  5  6  7


root1 = new Node(1);
root1.left = new Node(2);
root1.right = new Node(3);
root1.left.left = new Node(4);
root1.left.right = new Node(5);
root1.right.left = new Node(6);
root1.right.right = new Node(7);


//              2
//            4   5

root2 = new Node(3);
root2.left = new Node(6);
root2.right = new Node(7);

console.log(checkSubtree(root1, root2));


function inOrder(node, list){
    if(node == null)
        return;

    //left
    inOrder(node.left, list);
    //node
    list.push(node.data);
    //right
    inOrder(node.right, list);
}

function preOrder(node, list){
    if(node == null)
        return;

    //node
    list.push(node.data);
    //left
    preOrder(node.left, list);
    //right
    preOrder(node.right, list);
}

function checkSubtree(tree, sub){
    if(sub == null)
        return true;
    if(tree == null)
        return false;


    //inOrder
    let first = [];
    let second = [];
    inOrder(tree, first);
    inOrder(sub, second);

    if(!listToString(first).includes(listToString(second)))
        return false;

    //preOrder
    first = [];
    second = [];
    preOrder(tree, first);
    preOrder(sub, second);

    if(!listToString(first).includes(listToString(second)))
        return false;

    return true;
}

function listToString(list){
    return list.toString().replace(/,/g,'');
}

File: recursion/binary-subtree/java/BinarySubtree.java
package com.basicsstrong.algorithm.recursion;

import java.util.ArrayList;
import java.util.List;

class Node
{
	int data;
	Node left = null, right = null;

	Node(int data) {
		this.data = data;
	}
}

public class BinarySubtree {


		public static void main(String[] args) {

		/*
		 * 			1 
		 * 		   / \ 
		 * 		  2   3 
		 * 		 / \ / \ 
		 * 		4  5 6  7
		 */
		
		//that's the tree, that we are constructing here.
		Node root = new Node(1);
		root.left = new Node(2);
		root.right = new Node(3);
		root.left.left = new Node(4);
		root.left.right = new Node(5);
		root.right.left = new Node(6);
		root.right.right = new Node(7);

		/*
		 * 			4 
		 * 		   / \ 
		 * 		  5   6 
		 */
		Node root2 = new Node(4);
		root2.left = new Node(5);
		root2.right = new Node(6);

		System.out.println(checkSubtree(root,root.left));	//true
		System.out.println(checkSubtree(root,root2));	//false
		
	}
	
	// Function to store in-order traversal of the tree in a list
	public static void inorder(Node node, List<Integer> list) {
		if (node == null) {
			return;
		}
		inorder(node.left, list);
		list.add(node.data);
		inorder(node.right, list);
	}

	// Function to store pre-order traversal of the tree in a list
	public static void preorder(Node node, List<Integer> list) {
		if (node == null) {
			return;
		}
		list.add(node.data);
		preorder(node.left, list);
		preorder(node.right, list);
	}

	//Create a function to check if given tree is subtree or not: 
	//as arguments, we'll pass given two trees.
	public static boolean checkSubtree(Node tree, Node sub) {

		//the case if tree sub is null, return true as subtree may be empty
		if (sub == null)
			return true;
		
		//and if tree is null, sub cannot be subtree of it - so return false
		if (tree == null)
			return false;

		//now we store in-order traversal of both trees in first and second respectively
		List<Integer> first = new ArrayList<>();
		inorder(tree, first);

		List<Integer> second = new ArrayList<>();
		inorder(sub, second);	
		
		//Now we return false if second list is not a sub-list of first list
		if(!listToString(first).contains(listToString(second)))
			return false;

		//otherwise we continue to check whether preOrder traversal of subtree is also subset of preOrder of tree
		
		//removing data from these lists.
		first.clear();
		second.clear();
		
		//and storing preOrder traversal into them
		preorder(tree, first);
		preorder(sub, second);
		
		//at last we return the result, that will be true if first contains second else false.
		return (listToString(first).contains(listToString(second)));
	}

	//utility method to convert list to string, by using list.toString and replacing brackets and commas
	private static String listToString(List<Integer> list) {
		return list.toString().replace("[", "").replace("]", "").replace(",", "");
	}
	
}

